klippa/src/variations.rs

```diff
@@ -0,0 +1,532 @@
+        let mut new_ri_count: u16 = 0;
+
+        for (r, delta_type) in delta_sz.iter().enumerate() {
+            if *delta_type == DeltaSize::Zero {
+                continue;
+            }
+
+            if *delta_type == DeltaSize::Word {
+                let new_r = word_index as usize;
+                word_index += 1;
+                ri_map[new_r] = r;
+            } else {
+                let new_r = non_word_index as usize;
+                non_word_index += 1;
+                ri_map[new_r] = r;
+            }
+            new_ri_count += 1;
+        }
+
+        let new_word_delta_count = if has_long {
+            new_word_count | 0x8000
+        } else {
+            new_word_count
+        };
+        s.embed(new_word_delta_count)?;
+        s.embed(new_ri_count)?;
+
+        let region_indices_pos = s.allocate_size(new_ri_count as usize * 2, false)?;
+        let src_region_indices = self.region_indexes();
+        for (idx, src_idx) in ri_map.iter().enumerate().take(new_ri_count as usize) {
+            let Some(old_r) = src_region_indices.get(*src_idx) else {
+                return Err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            };
+
+            let Some(region) = region_map.get(old_r.get() as u32) else {
+                return Err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            };
+            s.copy_assign(region_indices_pos + idx * 2, *region as u16);
+        }
+
+        let new_row_size = ItemVariationData::delta_row_len(new_word_delta_count, new_ri_count);
+        let new_delta_bytes_len = new_item_count as usize * new_row_size;
+
+        let delta_bytes_pos = s.allocate_size(new_delta_bytes_len, false)?;
+        for i in 0..new_item_count as usize {
+            let Some(old_i) = inner_map.backward(i as u32) else {
+                return Err(SerializeErrorFlags::SERIALIZE_ERROR_OTHER);
+            };
+
+            let old_i = *old_i as usize;
+            for (r, old_r) in ri_map.iter().enumerate().take(new_ri_count as usize) {
+                set_item_delta(
+                    s,
+                    delta_bytes_pos,
+                    i,
+                    r,
+                    get_item_delta(self, old_i, *old_r, src_row_size, src_delta_bytes),
+                    new_row_size,
+                    has_long,
+                    new_word_count as usize,
+                )?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn get_item_delta(
+    var_data: &ItemVariationData,
+    item: usize,
+    region: usize,
+    row_size: usize,
+    delta_bytes: &[u8],
+) -> i32 {
+    if item >= var_data.item_count() as usize || region >= var_data.region_index_count() as usize {
+        return 0;
+    }
+
+    let p = item * row_size;
+    let word_delta_count = var_data.word_delta_count();
+    let word_count = (word_delta_count & 0x7FFF) as usize;
+    let is_long = word_delta_count & 0x8000 != 0;
+
+    #[allow(clippy::collapsible_else_if)]
```

Comment:
In general if we're ignoring a lint I think it's worth a comment explaining why we think it's justified.

In this case I assume this is ported directly from the harfbuzz, and so you want to keep the translation as literal as possible?

If so I'd just say this, and include a link to the original.